symengine/nan.cpp

```diff
@@ -0,0 +1,56 @@
+#include <symengine/nan.h>
+
+namespace SymEngine
+{
+
+NaN::NaN()
+{
+    SYMENGINE_ASSIGN_TYPEID()
+}
+
+hash_t NaN::__hash__() const
+{
+    hash_t seed = NAN;
+    return seed;
+}
+
+//! NaN is mathematically not equal to anything else, even NaN itself.
+bool NaN::__eq__(const Basic &o) const
+{
+    return false;
+}
+
+int NaN::compare(const Basic &o) const
+{
+    if (is_same_type(*this, o))
```

Comment:
See https://github.com/symengine/symengine/blob/c1cd1adbf15d782bf48048b31428702bef81fbab/symengine/infinity.cpp#L68-L73 for how to implement this.